rfcs/text/0408-versioning-types.md

```diff
@@ -0,0 +1,344 @@
+
+[summary]: #summary
+
+The Type System outlined in [RFC 0352](https://github.com/blockprotocol/blockprotocol/blob/am/rfc-graph-type-system/rfcs/text/0352-graph-type-system.md) introduced classes of types that are used to describe the data used and produced by blocks and embedding applications. The types are possibly (ideally often) publicly accessible, and their hierarchy allows for them to be composed together into new types. Due to this, changes in a type can affect the structure of another one. This RFC defines how types are versioned, to mitigate potential problems that would arise without versioning.
+
+# Motivation
+
+[motivation]: #motivation
+
+The Type System is intended to allow for **reusable and shared** descriptions of data. As such, a version of a block should be able to tie itself to a _specific_ description of the structure of the data it requires as if types are able to be updated independently of blocks then this would cause problems. Additionally, if someone (Alice) creates an entity type that refers to someone else's (Bob's) "address" property type, we do not want Bob to be able to change what Alice meant in her entity type by arbitrarily and implicitly updating it.
+
+Being able to refer to specific, immutable versions of types should mitigate or solve the majority of the issues stemming from this.
+
+This RFC proposes that every Type (i.e. Data Type, Property Type, Entity Type, or Link Type) should be associated with a unique version number. This version number will be part of the URI that is used to uniquely identify the specific instance of the Type, to access the Type's schema, and to refer to the Type from other Types.
+
+We also outline a method to compare compatibility between types, and use this in the [Rationale and Alternatives](#rationale-and-alternatives) section to explain how this replaces the need for some other versioning scheme such as semantic versioning.
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+## Type Versions
+
+The _version number_ of a type is an incrementing positive integer that increases by 1 with each new version, starting from 1.
+
+## Type URIs
+
+The _base URI_ of a type is a _unique_ identifier for the type irrespective of its version.
+
+The _versioned URI_ of a type is made up of:
+
+```json
+"${base_uri}/v/${version_number}"
+```
+
+We opt not to constrain the format of the `base_uri` to enable flexibility in non-public use-cases or in domains with constraints that we cannot predict. Despite this, we _suggest_ that the base URIs should have a consistent format for all types hosted on your domain, and should generally be humanly readable in a way that describes your type. For example, one may pick:
+
+```json
+"http://yourdomain/namespace/property-type/type_name"
+```
+
+as the identifier (assuming that the domain decides that type names are unique across a namespace).
+
+Where it's possible to guarantee, this URI should be valid indefinitely.
```

Comment:
For my understanding, would this recommendation mean that if a user were to create a `Person` entity type, delete it, and create another `Person` type down the line, that they will have unintentionally prevented themselves from being able to re-use the `http://yourdomain/namespace/entity-type/Person` URI ever again? Or would the version counter not restart from `0`, and the re-created entity type be considered "newer versions" of the prior one? Or is this up to the implementation to decide

Reply:
Under the guidance of this proposal, the expectation is that publicly accessible URIs are permanent and immutable, that is, they cannot be deleted.

Now if people are operating within a closed sphere of their domain and wish to have things that aren't really _publicly accessible_ then we can't control how they want to use these.

In those circumstances I think they could decide themselves if the URI needs to be reserved in the future or not.

Reply:
I think that's something we do need to discuss but I'm not sure this PR is the right place for it (open to being corrected on that if we really do need to provision for this in the spec).

I think we can have these in the Github Discussions when this is merged, while we figure out our URI format for blockprotocol.org

Reply:
This doesn't block merging

Reply:
It may be worth considering allowing for version IDs to be UUIDs, in which case a rule such as "all _versioned_ URIs must be valid indefinitely" will be a lot easier for us to adhere to in a world where users can acquire previously used namespaces, or re-use previous _non-versioned_ URIs.

Reply:
My main concern was that users will end up with `http://.../Person-3` in their namespace after deleting the (public) entity types `http://.../Person` and `http://.../Person-2`, and there's not a lot we can do about it in the outlined version system. But I'm realising if we don't support deleting public types anyway, then this won't really be an issue as the user will be forced to create a new version of the `Person` type if they wish to change it.

Reply:
> It may be worth considering allowing for version IDs to be UUIDs, in which case a rule such as "all versioned URIs must be valid indefinitely" will be a lot easier for us to adhere to in a world where users can acquire previously used namespaces, or re-use previous non-versioned URIs.

I don't think we should maintain URLs with namespaces pointing to a resource created by a user, when a different user now owns that namespace (and has other resources at URLs using it). If we want URIs to be valid indefinitely, I think once someone has used a namespace it can't be used by anyone else.

Reply:
I also see the value in UUIDs but I don't think it outweighs the value of the current approach which @nonparibus identified, where we can programmatically discover later (and earlier) versions by simply incrementing the counter.